File: examples/ruby/example_test.rb
#!/use/bin/enc ruby
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

require 'test/unit'
require 'qpid_proton'
require 'socket'

def random_port
  return Random.new.rand(10000) + 10000
end

def wait_port port
  # Wait for something to be listening on a port.
  for i in 0..100
    begin
      TCPSocket.open("", port).close
      return
    rescue Errno::ECONNREFUSED
      sleep(0.1)
    end
  end
  raise "timeout waiting for port #{port}"
end

module ExampleTest

  def run_script(script, port)
    assert File.exist? script
    cmd = [RbConfig.ruby, script]
    cmd += ["-a", ":#{port}/examples"] if port
    return IO.popen(cmd)
  end


  def assert_output(script, want, port=nil)
    out = run_script(script, port)
    assert_equal want.strip, out.read.strip
  end

  def test_helloworld
    assert_output("#{@dir}/helloworld.rb", "Hello world!", $broker_port)
  end

  def test_simple_send_recv
    assert_output("#{@dir}/simple_send.rb", "All 100 messages confirmed!", $broker_port)
    want = (0..99).reduce("") { |x,y| x << "Received: sequence #{y}\n" }
    assert_output("#{@dir}/simple_recv.rb", want, $broker_port)
  end

  def test_simple_send_direct_recv
    port = random_port
    recv = run_script("#{@dir}/direct_recv.rb", port)
    sleep(0.1)                  # FIXME aconway 2016-01-05: wait for ready.
    assert_output("#{@dir}/simple_send.rb", "All 100 messages confirmed!", port)
    Process.kill :TERM, recv.pid
    want = (0..99).reduce("") { |x,y| x << "Received: sequence #{y}\n" }
    assert_equal want.strip, recv.read.strip
  end

  def test_direct_send_simple_recv
    port = random_port
    send = run_script("#{@dir}/direct_send.rb", port)
    sleep(0.1)                  # FIXME aconway 2016-01-05: wait for ready.
    want = (0..99).reduce("") { |x,y| x << "Received: sequence #{y}\n" }
    assert_output("#{@dir}/simple_recv.rb", want, port)
    Process.kill :TERM, send.pid
    assert_equal "All 100 messages confirmed!", send.read.strip
  end

  def test_client_server
    want =  <<EOS
-> Twas brillig, and the slithy toves
<- TWAS BRILLIG, AND THE SLITHY TOVES
-> Did gire and gymble in the wabe.
<- DID GIRE AND GYMBLE IN THE WABE.
-> All mimsy were the borogroves,
<- ALL MIMSY WERE THE BOROGROVES,
-> And the mome raths outgrabe.
<- AND THE MOME RATHS OUTGRABE.
EOS
    srv = run_script("#{@dir}/server.rb", $broker_port)
    assert_output("#{@dir}/client.rb", want.strip, $broker_port)

  ensure
    Process.kill :TERM, srv.pid if srv
  end

end

def start_broker dir
  $broker_port = random_port
  $broker = spawn("#{RbConfig.ruby} #{dir}/broker.rb -a :#{$broker_port}")
  wait_port $broker_port
end

def stop_broker
  Process.kill :TERM, $broker
  Process.wait $broker
end


class ReactorTest < Test::Unit::TestCase
  DIR = "reactor"
  class << self
    def startup; start_broker DIR end
    def shutdown; stop_broker end
  end
  def setup; @dir = DIR end
  include ExampleTest
end

class EngineTest < Test::Unit::TestCase
  DIR = "engine"
  class << self
    def startup; start_broker DIR end
    def shutdown; stop_broker end
  end
  def setup; @dir = DIR end
  include ExampleTest
end

File: examples/exampletest.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License
#

# A test library to make it easy to run unittest tests that start,
# monitor, and report output from sub-processes. In particular
# it helps with starting processes that listen on random ports.

import unittest
import os, sys, socket, time, re, inspect, errno, threading
from  random import randrange
from subprocess import Popen, PIPE, STDOUT
from copy import copy
import platform
from os.path import dirname as dirname

def bind0():
    """Bind a socket with bind(0) and SO_REUSEADDR to get a free port to listen on"""
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)    
    sock.bind(('', 0))
    return sock

# Monkeypatch add port() and with support to socket.socket
socket.socket.port = lambda(self): socket.getnameinfo(self.getsockname(), 0)[1]
socket.socket.__enter__ = lambda(self): self
def socket__exit__(self, *args): self.close()
socket.socket.__exit__ = socket__exit__

class ProcError(Exception):
    """An exception that captures failed process output"""
    def __init__(self, proc, what="bad exit status"):
        out = proc.out.strip()
        if out:
            out = "\nvvvvvvvvvvvvvvvv\n%s\n^^^^^^^^^^^^^^^^\n" % out
        else:
            out = ", no output)"
        super(Exception, self, ).__init__(
            "%s %s, code=%s%s" % (proc.args, what, proc.returncode, out))

class NotFoundError(ProcError):
    pass

class Proc(Popen):
    """A example process that stores its output, optionally run with valgrind."""

    if "VALGRIND" in os.environ and os.environ["VALGRIND"]:
        env_args = [os.environ["VALGRIND"], "--error-exitcode=42", "--quiet", "--leak-check=full"]
    else:
        env_args = []

    @property
    def out(self):
        self._out.seek(0)
        return self._out.read()

    def __init__(self, args, **kwargs):
        """Start an example process"""
        args = list(args)
        self.args = args
        self.kwargs = kwargs
        self._out = os.tmpfile()
        try:
            Popen.__init__(self, self.env_args + self.args, stdout=self._out, stderr=STDOUT, **kwargs)
        except OSError, e:
            if e.errno == errno.ENOENT:
                raise NotFoundError(self, str(e))
            raise ProcError(self, str(e))
        except Exception, e:
            raise ProcError(self, str(e))

    def kill(self):
        try:
            if self.poll() is None:
                Popen.kill(self)
        except:
            pass                # Already exited.
        return self.out

    def wait_out(self, timeout=10, expect=0):
        """Wait for process to exit, return output. Raise ProcError  on failure."""
        t = threading.Thread(target=self.wait)
        t.start()
        t.join(timeout)
        if self.poll() is None:      # Still running
            self.kill()
            raise ProcError(self, "still running after %ss" % timeout)
        if expect is not None and self.poll() != expect:
            raise ProcError(self)
        return self.out

    def wait_re(self, regexp, timeout=10):
        """
        Wait for regexp to appear in the output, returns the re.search match result.
        The target process should flush() important output to ensure it appears.
        """
        if timeout:
            deadline = time.time() + timeout
        while timeout is None or time.time() < deadline:
            match = re.search(regexp, self.out)
            if match:
                return match
            time.sleep(0.01)    # Not very efficient
        raise ProcError(self, "gave up waiting for '%s' after %ss" % (regexp, timeout))

# Work-around older python unittest that lacks setUpClass.
if hasattr(unittest.TestCase, 'setUpClass') and  hasattr(unittest.TestCase, 'tearDownClass'):
    TestCase = unittest.TestCase
else:
    class TestCase(unittest.TestCase):
        """
        Roughly provides setUpClass and tearDownClass functionality for older python
        versions in our test scenarios. If subclasses override setUp or tearDown
        they *must* call the superclass.
        """
        def setUp(self):
            if not hasattr(type(self), '_setup_class_count'):
                type(self)._setup_class_count = len(
                    inspect.getmembers(
                        type(self),
                        predicate=lambda(m): inspect.ismethod(m) and m.__name__.startswith('test_')))
                type(self).setUpClass()

        def tearDown(self):
            self.assertTrue(self._setup_class_count > 0)
            self._setup_class_count -=  1
            if self._setup_class_count == 0:
                type(self).tearDownClass()

class ExampleTestCase(TestCase):
    """TestCase that manages started processes"""
    def setUp(self):
        super(ExampleTestCase, self).setUp()
        self.procs = []

    def tearDown(self):
        for p in self.procs:
            p.kill()
        super(ExampleTestCase, self).tearDown()

    def proc(self, *args, **kwargs):
        p = Proc(*args, **kwargs)
        self.procs.append(p)
        return p

class BrokerTestCase(ExampleTestCase):
    """
    ExampleTest that starts a broker in setUpClass and kills it in tearDownClass.
    Subclass must set `broker_exe` class variable with the name of the broker executable.
    """

    @classmethod
    def setUpClass(cls):
        sock = bind0()
        cls.port = sock.port()
        cls.addr = "127.0.0.1:%s/examples" % (cls.port)
        cls.broker = None       # In case Proc throws, create the attribute.
        cls.broker = Proc(cls.broker_exe + ["-a", cls.addr], bufsize=0)
        try:
            cls.broker.wait_re("listening")
        except Exception, e:
            cls.broker.kill()
            raise
        finally:
            sock.close()

    @classmethod
    def tearDownClass(cls):
        if cls.broker: cls.broker.kill()

    def tearDown(self):
        b = type(self).broker
        if b and b.poll() !=  None: # Broker crashed
            type(self).setUpClass() # Start another for the next test.
            raise ProcError(b, "broker crash")
        super(BrokerTestCase, self).tearDown()

if __name__ == "__main__":
    unittest.main()
